cpu/native: add periph_eeprom driver implementation

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This implementation is an emulation of an EEPROM for native using an in memory buffer modified during runtime. This buffer is filled from a file on disk at instance startup and dumped to this file at exit/reboot.
This provides the same persistence mechanism as a hardware eeprom.

The eeprom file is provided via termflags. I'm not 100% sure the integration is correct here, so comments are welcome.
The EEPROM size for native is 1024kB.

The periph_eeprom test application is modified:
- native board is now the default
- the test script triggers a reboot of the application and checks that previous data are still available on EEPROM

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

The following commands are working (now the first one is building for native):
```
$ make -C tests/periph_eeprom flash test
$ make BOARD=nucleo-l073rz -C tests/periph_eeprom flash test
$ make BOARD=nucleo-l152re -C tests/periph_eeprom flash test
```

Another thing that could be tested on native is the `EEPROM_FILE` variable:
```
$ ls /tmp/riot_test.eeprom 
ls: cannot access '/tmp/riot_test.eeprom': No such file or directory
$ EEPROM_FILE=/tmp/riot_test.eeprom make -C tests/periph_eeprom all test
$ hexdump /tmp/riot_test.eeprom 
0000000 4141 4141 0000 0000 0000 0000 0000 0000
0000010 0000 0000 0000 0000 0000 0000 0000 0000
*
0000200 0041 0000 0000 0000 0000 0000 0000 0000
0000210 0000 0000 0000 0000 0000 0000 0000 0000
*
00003f0 0000 0000 0000 0000 0000 0000 4141 4141
0000400
```
=> the file is automatically created and content is correct: the last tests are writing AAAA at the beginning and at the end of the EEPROM.

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

None

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

fjmolinas commented on cpu/native/periph/eeprom.c:
```diff
@@ -0,0 +1,102 @@
+    fclose(fp);
+}
+
+void eeprom_native_write(void)
```
This is probably not the best simulation of an eeprom since you are always overwriting the whole file, but I guess it doesn't matter in native that you always have a complete copy in RAM.

fjmolinas commented on cpu/native/include/periph_cpu.h:
```diff
@@ -91,6 +91,13 @@ typedef enum {
+ * @name    EEPROM configuration
+ * @{
+ */
+#define EEPROM_SIZE             (1024U)  /* 1kB */
```
Maybe this could be configurable?

aabadie replied:
Done

fjmolinas commented on cpu/native/include/eeprom_native.h:
```diff
@@ -0,0 +1,58 @@
+/**
+ * @brief   Write the EEPROM content to the configured file
+ *
+ * This function is called once before rebooting or before exiting the native
```
The issue with this is that if it hardfaults all intermediate writes would not be completed, which would be the case on a device eeprom.

aabadie replied:
This is fixed

aabadie commented on cpu/native/startup.c:
```diff
@@ -312,6 +319,12 @@ void usage_exit(int status)
+#endif
+#ifdef MODULE_PERIPH_EEPROM
+    real_printf(
+"    -M <eeprom> , --eeprom=<eeprom>\n"
```
for memory. -e, -E were already taken for other things (stderr).

fjmolinas commented on cpu/native/periph/eeprom.c:
```diff
@@ -0,0 +1,105 @@
+
+static inline void eeprom_native_write(void)
+{
+    FILE *fp = fopen(eeprom_file, "w");
```
```suggestion
    FILE *fp = real_fopen(eeprom_file, "w");
```

fjmolinas commented on cpu/native/periph/eeprom.c:
```diff
@@ -0,0 +1,105 @@
+
+void eeprom_native_read(void)
+{
+    FILE *fp = fopen(eeprom_file, "r");
```
```suggestion
    FILE *fp = real_fopen(eeprom_file, "r");
```

fjmolinas commented on tests/periph_eeprom/tests/01-run.py:
```diff
@@ -11,9 +11,21 @@
 
 
 def testfunc(child):
+    # make sure the shell is ready before sending the test command
+    child.expect_exact('>')
```
Please remove this, already matched by `shell_interactive_sync`.

fjmolinas commented on tests/periph_eeprom/tests/01-run.py:
```diff
@@ -11,9 +11,21 @@
     child.expect('SUCCESS')
 
+    # reboot the device to ensure data are still available on EEPROM
+    child.sendline('reboot')
```
This will require blacklisting boards using `stdio_cdc_acm`

aabadie replied:
Indeed, but `stdio_cdc_acm` is not a feature that could be blacklisted. I could at minimum blacklist `bootloader_arduino` or `bootloader_nrfutil` but that doesn't cover all boards using `stdio_cdc_acm`.
For the moment, the good thing is (I think) that there's no board providing `periph_eeprom` and using `stdio_cdc_acm`.

fjmolinas replied:
Lets leave it like that for now, this is not the only test with that issue.

aabadie commented:
Apart from https://github.com/RIOT-OS/RIOT/pull/12709/files/01bf4471411d7f0ea5fd911fcb3f846b573490f9#diff-9603074c5dc13cfbe17dcb106e55f20d, I think I addressed all your comments @fjmolinas

fjmolinas commented:
@aabadie please squash and trigger murdock!

aabadie commented:
Squashed!

aabadie commented:
All green here @fjmolinas :)
